### PR TITLE
Preserve data

### PR DESCRIPTION
**Pull request by zkr024**
* Create JSON files for books, persons, and rentals
* Load data from the JSON files when the app begins.
* Create a separate folder for the "loadData" functionality
* Store data recorded from books, persons, and rentals added
* The Store data functionality triggers when closing the app.

image

### CODE REVIEW

**Review by Davidosky007 — COMMENTED** [review]
Hi @zkr024 @Hadi111jafari 👋, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, that wasnt able to run, i kept getting errors even after following setup instructions in the readme file. below ⬇️ is a scrrenshoot of the error i get 
image


Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

**Comment by zkr024** [comment]
> Hi @zkr024 @Hadi111jafari 👋,
> 
> While you made a great effort in this project, unfortunately, I cannot proceed to review your code.
> 
> ## Invalid Code Review Request
> You have submitted a project, that wasnt able to run, i kept getting errors even after following setup instructions in the readme file. below ⬇️ is a scrrenshoot of the error i get image
> 
> Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏
> 
> Cheers and Happy coding!👏👏👏
> 
> Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
> 
> **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**
> 
> _Invalid Code Review Request **does not count** into the code reviews limit._

**Comment by zkr024** [comment]
@Davidosky007 Thanks for the heads up...I'm not sure what could have happened. I will update the readme file and check the project to see why it is not loading.

File: loadData/book_data.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  return if @file_book.empty?
+
+  data_book = JSON.parse(@file_book)
+  data_book.map do |value|
+    title = value['title']
+    author = value['author']
+    book_data = Book.new(title, author)
+    book << book_data
+  end
+end
```
Comment (andersonlebon) [inline loadData/book_data.rb]: image

- [x] Great job on saving data in JSON files, However, in the requirements it said that you should handle the case where one of the JSON files is missing the application should not throw an error. I deleted the `/data/book.json` file and the app crashed. Check this hint solution :point_down: 

```rb
# Kindly note that this is just an example solution, feel free to solve it the way you want, just make sure the app does nit crash when this file is missing
require 'json'
require_relative '../school/book'


def book_data(book)
  File.write('./data/book.json', JSON.pretty_generate([])) unless File.exists?('./data/book.json')
  @file_book = File.read('./data/book.json')
  return if @file_book.empty?

  data_book = JSON.parse(@file_book)
  data_book.map do |value|
    title = value['title']
    author = value['author']
    book_data = Book.new(title, author)
    book << book_data
  end
end
```

File: loadData/person_data.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      specialization = value['specialization']
+      person_data = Teacher.new(specialization, age, name)
+    else
+      classroom = value['classroom']
+      person_data = Student.new(classroom, age, name)
+    end
+    person_data.id = value['id']
+    person << person_data
+  end
+end
```
Comment (andersonlebon) [inline loadData/person_data.rb]: image

- [x] Great job on saving data in JSON files, However, in the requirements it said that you should handle the case where one of the JSON files is missing the application should not throw an error. I deleted the `/data/person.json` file and the app crashed. Check this hint solution :point_down: 

```rb
# Kindly note that this is just an example solution, feel free to solve it the way you want, just make sure the app does not  crash when this file is missing

require 'json'
require_relative '../school/student'
require_relative '../school/teacher'


def person_data(person)
  File.write('./data/person.json', JSON.pretty_generate([])) unless File.exists?('./data/person.json')
  @file_person = File.read('./data/person.json')
  return if @file_person.empty?

  data_person = JSON.parse(@file_person)
  data_person.map do |value|
    name = value['name']
    age = value['age']
    if value['classroom'].nil?
      specialization = value['specialization']
      person_data = Teacher.new(specialization, age, name)
    else
      classroom = value['classroom']
      person_data = Student.new(classroom, age, name)
    end
    person_data.id = value['id']
    person << person_data
  end
end

File: loadData/rental_data.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    name = value['Name']
+    book = value['Book']
+    author = value['Author']
+    date = value['Date']
+    book_info = Book.new(book, author)
+    person_info = Person.new(age, name)
+    rental_info = Rental.new(date, book_info, person_info)
+    rental << rental_info
+  end
+end
```
Comment (andersonlebon) [inline loadData/rental_data.rb]: image

- [x] Great job on saving data in JSON files, However, in the requirements it said that you should handle the case where one of the JSON files is missing the application should not throw an error. I deleted the `/data/rental.json` file and the app crashed. Check this hint solution :point_down: 

```rb
# Kindly note that this is just an example solution, feel free to solve it the way you want, just make sure the app does not  crash when this file is missing

def rental_data(rental)
  File.write('./data/rental.json', JSON.pretty_generate([])) unless File.exists?('./data/rental.json')
  @file_rental = File.read('./data/rental.json') 
  return if @file_rental.empty?

  data_rental = JSON.parse(@file_rental)
  data_rental.map do |value|
    age = value['Age']
    name = value['Name']
    book = value['Book']
    author = value['Author']
    date = value['Date']
    book_info = Book.new(book, author)
    person_info = Person.new(age, name)
    rental_info = Rental.new(date, book_info, person_info)
    rental << rental_info
  end
end

```

**Review by andersonlebon — CHANGES_REQUESTED** [review]
## Hi @zkr024 

Good job so far! I know you have worked hard on making this milestone But  There are some issues that you still need to work on to go to the next project but you are almost there!

almost-there
## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- Good commit history :heavy_check_mark: 
- Good PR description :heavy_check_mark: 
## Required Changes ♻️

- [ ] Kindly check comments under the review.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: loadData/book_data.rb (hunk 1 above)
Comment (zkr024) [reply loadData/book_data.rb]: @andersonlebon Thank you for the feedback, I will update the project.

**Review by Cathella — APPROVED** [review]
## 🔥 Status: Approved 🟢

### Hi @zkr024 👋🏻, 

Good job on making the requested changes 💪🏻!
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 💪🏻

✅ Missing json file case handled 👍🏻 


Cheers and Happy coding!👏👏👏

rosycheeks-mochi-peach


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**